aioxmpp/ibr/service.py

```diff
@@ -0,0 +1,253 @@
+                setattr(iq.payload, key, value)
+        yield from client.send(iq)
+
+    @asyncio.coroutine
+    def cancel_registration(self, client):
+        """
+        Cancels the currents client's account with the server.
+
+        :param client: Client who wants to cancel.
+        :type client: :class:`aioxmpp.Client`
+
+        :return:`None`
+
+        Even if the cancelation is succesful, this method will raise an
+        exception due to he account no longer exists for the server, so the
+        client will fail.
+        To continue with the execution, this method should be surrounded by a
+        try/except statement.
+        """
+        iq = aioxmpp.IQ(
+            to=client.local_jid.bare().replace(localpart=None),
+            type_=aioxmpp.IQType.SET,
+            payload=xso.Query()
+        )
+
+        iq.payload.remove = True
+        yield from client.send(iq)
+
+    @staticmethod
+    @asyncio.coroutine
+    def connect_and_send(iq, loop, metadata, peer_jid, timeout):
+        options = list((
+            yield from aioxmpp.discover_connectors(
+                peer_jid.domain, loop=loop
+            )
+        ))
+        for host, port, conn in options:
+            transport, xmlstream, features = yield from conn.connect(
+                loop,
+                metadata,
+                peer_jid.domain,
+                host,
+                port,
+                timeout,
+                base_logger=logger
+            )
+            reply = yield from aioxmpp.send_and_wait_for(
+                xmlstream, [iq], [aioxmpp.IQ]
+            )
+        return reply
+
+    @asyncio.coroutine
+    def get_registration_fields(self,
+                                peer_jid,
+                                timeout=60,
+                                no_verify=True
+                                ):
+        """
+        A query is sent to the server to obtain the fields that need to be
+        filled to register with the server.
+
+        :param peer_jid: Server where he query will be sent.
+        :type peer_jid: :class:`aioxmpp.JID`
+
+        :param timeout: Maximum time in seconds to wait for an IQ response, or
+                        :data:`None` to disable the timeout.
+        :type timeout: :class:`~numbers.Real` or :data:`None`
+
+        :param no_verify: Specifies whether the server's TLS is verified.
+        :type no_verify: :class:`aioxmpp.JID`
+
+        :return: :attr:`list`
+        """
+        iq = aioxmpp.IQ(
+            to=peer_jid.bare().replace(localpart=None),
+            type_=aioxmpp.IQType.GET,
+            payload=xso.Query()
+        )
+        iq.autoset_id()
+
+        loop = asyncio.get_event_loop()
+        metadata = aioxmpp.make_security_layer('', no_verify=no_verify)
+
+        try:
+            reply = yield from self.connect_and_send(iq, loop, metadata, peer_jid, timeout)
+
+            return [a for a in dir(reply.payload)
```

Comment:
Perhaps simply return the payload XSO instead instead of decomposing it into a list. If the payload XSO grows a public method later on things get uncomfortable with this code. (The person adding the method might overlook this code and then the method will appear in the dictionary: not what we want).

Reply:
@sebastianriese is right that it’s better to return the XSO payload here instead of trying to parse it in the method. Try to keep methods as minimalistic as possible (without over-doing it), this helps with re-usability and testability.

I suggest to add a method (e.g. ``get_used_fields`` or something similar) to the ``Query`` class instead which returns the set of child element names which were given. This can be done for example by iterating over ``Query.CHILD_MAP``. Try this:

```
q = aioxmpp.ibr.xso.Query()
q.username = ""
q.password = ""
print([
    tag
    for tag, descriptor in q.CHILD_MAP.items()
    if descriptor.__get__(q, type(q)) is not None
])
```

This gives you a list of tags of all (child) properties which are not None. You will want to exclude the instructions field from the result of that method.